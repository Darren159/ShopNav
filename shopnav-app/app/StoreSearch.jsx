import React, { useState, useEffect } from 'react';
import { Text, View, ActivityIndicator, FlatList, StyleSheet, Image, TouchableOpacity } from 'react-native';
import SearchBar from '../../Components/SearchBar';
import { SafeAreaView } from 'react-native';
import { db } from '../../firebaseConfig';
import { collection, getDocs } from 'firebase/firestore';
import filter from 'lodash.filter';
import { useRouter } from 'expo-router';

export default function StoreSearch() {

    //const  navigation = useNavigation();    
    //for filtering search function
    const handleSearch = (query) => {
        if (query) {
            const formattedQuery = query.toLowerCase();
            const filteredData = filter(fullData, (item) => {
            return contains(item, formattedQuery);
            });
            setData(filteredData);

            //console.log(`Searching for ${query}`);
        } else {
            //If query is cleared, reset the data to the full list
            setData(fullData);
        }
       
    };

    const contains = ({adjacent} , query) => {
        if (adjacent[0].includes(query)) {
            return true;
        } else {
            return false;
        }
    };

    //for handling pressing of location
    const router = useRouter();

    //for pop out of details page
    const pressHandler = async (location) => {
        
        router.push({ pathname: '/PlaceDetailsScreen', params: { locName: location } 
            //name: result.name, 
            //formatted_phone_number: result.formatted_phone_number,
            //opening_hours: result.opening_hours,
            //photos: result.photos[0].photo_reference,
            //business_status: result.business_status,
            //rating: result.rating,
            //reviews: result.reviews,
            //formatted_address: result.formatted_address, }, 
    })
        
    };

    
    
    const [isLoading, setIsLoading] = useState(false);
    const [data,setData] = useState([]);
    const [error, setError] = useState(null);
    const [fullData, setFullData] = useState([]);

    useEffect(() => {
        setIsLoading(true);
        fetchData();
        
    }, []);

    const fetchData = async () => {
            try {
                const colRef = collection(db, 'Westgate');

                //await until data is fetched
                const storeListSnapShot = await getDocs(colRef)

                const storeList = storeListSnapShot.docs.map(doc => doc.data())
                
                //sorting the list by alphabets
                storeList.sort((a, b) => a.adjacent[0].toLowerCase().localeCompare(b.adjacent[0].toLowerCase()));

                //filter out the unnecessary nodes
                const filteredStoreList = storeList.filter(item => !item.adjacent[0].toLowerCase().includes('aesop'));

                setFullData(filteredStoreList);

                //set data to full data set initially
                setData(filteredStoreList);

                setIsLoading(false);
                
            } catch (err) {
                console.error("Error fetching data: ", err);
                setError(err);
                setIsLoading(false);
            }
        };
    
    //loading interface
    if (isLoading) {
        return (
            <View style={{flex:1, justifyContent:"center", alignItems:"center" }}>
                <ActivityIndicator size={"large"} color="#5500dc" />
            </View>
        );
    }

    //error interface
    if ( error ) {
        return (
            <View style={{flex:1, justifyContent:"center", alignItems:"center" }}>
                <Text>Error in fetching data ... Please check your internet connection!</Text>
            </View>
        );
    }

        

    return (
        <SafeAreaView style= {styles.mainContainer}>
            <View style = {{ paddingBottom: 70}}>
                <Text>Store Search</Text>
                <SearchBar onSearch={handleSearch} />
                
                <FlatList
                data = {data}
                keyExtractor = {(item, index) => index.toString()}
                renderItem = {({item}) => (

                    
                    <TouchableOpacity onPress = { () => pressHandler(item.adjacent[0]) } >
                    <View style = {styles.itemContainer}>
                        <Image source = {{uri: "https://th.bing.com/th/id/R.99e2381b445a56f30bc79cb4b7d239a5?rik=VZlfhVCimbQLvA&riu=http%3a%2f%2fzenandtheartoftravel.com%2fwp-content%2fuploads%2f2012%2f06%2fDSC_0258.jpg&ehk=Q89%2bz8DrgosP78Myqq4Mga2oaT2T8bpmgQqhkHjUgMY%3d&risl=&pid=ImgRaw&r=0"}}
                        style = {styles.image} />
                        <Text style = {styles.textName}>{item.adjacent[0]}</Text>
                        {console.log(JSON.stringify(item.adjacent[0]))}

                    </View>
                    </TouchableOpacity>
                )}
                />
            </View>
        </SafeAreaView>
    );
}

const styles = StyleSheet.create({
    mainContainer: {
        flex: 1, 
        marginHorizontal: 20,
    },
    itemContainer: {
        flexDirection: "row",
        alignItems: "center",
        padding: 10,
        marginLeft: 0,
        marginTop: 20,
        borderRadius: 10,
        borderWidth: 1,
    },
    textName: {
        flex: 1,
        fontSize: 17,
        marginLeft: 10,
        fontWeight: "600",

    },
    image:  {
        marginLeft: 10,
        marginTop: 5,
        marginBottom: 5,
        width: 50,
        height: 50,
        borderRadius: 5,
    }
})