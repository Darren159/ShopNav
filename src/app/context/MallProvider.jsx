import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import PropTypes from "prop-types";
import fetchMalls from "../../services/fetchMalls";

export const MallContext = createContext();

export function MallProvider({ children }) {
  const [malls, setMalls] = useState([]);
  const [currentMall, setCurrentMall] = useState(null);

  useEffect(() => {
    const initializeMalls = async () => {
      try {
        const mallList = await fetchMalls();
        setMalls(mallList);
        setCurrentMall(mallList[0]); // Set the first mall as the current mall
      } catch (error) {
        Alert.alert("Error", error.message, [{ text: "OK" }], {
          cancelable: false,
        });
      }
    };
    initializeMalls();
  }, []);
  const value = useMemo(
    () => ({ malls, currentMall, setCurrentMall }),
    [malls, currentMall]
  );

  return <MallContext.Provider value={value}>{children}</MallContext.Provider>;
}

MallProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
