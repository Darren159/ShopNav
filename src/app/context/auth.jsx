import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import {
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import PropTypes from "prop-types";

import { auth } from "../../../firebaseConfig";

export const AuthContext = createContext({});

export function AuthProvider({ children }) {
  const [user, setUser] = useState(null);

  useEffect(() => {
    onAuthStateChanged(auth, (userState) => {
      if (userState) {
        setUser(userState);
      } else {
        setUser(null);
      }
    });
  }, []);

  const authContextValue = useMemo(
    () => ({
      user,
      setUser,
      signin: async (email, password) => {
        try {
          await signInWithEmailAndPassword(auth, email, password);
        } catch (error) {
          let errorMessage;
          
          switch (error.code) {
            case 'auth/user-not-found':
              errorMessage = 'There is no user record corresponding to this identifier. The user may have been deleted.';
              break;
            case 'auth/wrong-password':
              errorMessage = 'The password is invalid or the user does not have a password.';
              break;
            case 'auth/invalid-email':
              errorMessage = 'The email address is badly formatted.';
              break;
            default:
              errorMessage = 'Something went wrong. Please try again later.';
          }
      
          Alert.alert(
            "Authentication Failed",
            errorMessage,
            [
              { text: "OK" }
            ]
          );
        }
      },
      signout: async () => {
        try {
          await signOut(auth);
        } catch (e) {
          console.error(e);
        }
      },
    }),
    [user]
  );

  return (
    <AuthContext.Provider value={authContextValue}>
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
