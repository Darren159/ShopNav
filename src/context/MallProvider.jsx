import { createContext, useState, useEffect, useMemo } from "react";
import PropTypes from "prop-types";
import { collection, getDocs } from "firebase/firestore";
import { db } from "../../firebaseConfig";

export const MallContext = createContext();

export function MallProvider({ children }) {
  const [malls, setMalls] = useState([]);
  const [currentMall, setCurrentMall] = useState(null);

  useEffect(() => {
    const fetchMalls = async () => {
      const mallCollection = collection(db, "malls");
      const mallSnapshot = await getDocs(mallCollection);
      const mallList = mallSnapshot.docs.map((mall) => mall.id);
      setMalls(mallList);
      setCurrentMall(mallList[0]); // Set the first mall as the current mall
    };
    fetchMalls();
  }, []);
  const value = useMemo(
    () => ({ malls, currentMall, setCurrentMall }),
    [malls, currentMall]
  );

  return <MallContext.Provider value={value}>{children}</MallContext.Provider>;
}

MallProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
