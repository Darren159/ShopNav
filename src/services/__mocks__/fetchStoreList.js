const mockFetchStoreList = jest.fn();

export default mockFetchStoreList;
