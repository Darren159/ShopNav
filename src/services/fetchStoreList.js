import { collection, getDocs } from "firebase/firestore";
import { db } from "../../firebaseConfig";

export default async function fetchStoreList(currentMall) {
  const storeCollection = collection(db, "malls", currentMall, "stores");
  const storeSnapshot = await getDocs(storeCollection);
  if (storeSnapshot) {
    const storeList = storeSnapshot.docs.map((storeDoc) => ({
      id: storeDoc.id,
      ...storeDoc.data(),
    }));

    return storeList;
  }
  throw new Error("No stores found");
}
