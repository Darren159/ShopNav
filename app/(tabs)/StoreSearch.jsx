import { useState, useEffect } from 'react';
import { Text, View, ActivityIndicator, FlatList, StyleSheet, Image, TouchableOpacity, SafeAreaView } from 'react-native';
import { collection, getDocs } from 'firebase/firestore';
import filter from 'lodash.filter';
import { useRouter } from 'expo-router';
import { db } from '../../firebaseConfig';
import SearchBar from '../../Components/SearchBar';
import { useMalls } from '../../services/databaseService';
import MallPicker from '../../Components/MallPicker';

export default function StoreSearch() {

    // mall picker
    const { malls, currentMall, setCurrentMall } = useMalls();
    

    // const  navigation = useNavigation();    
    // for filtering search function
    const handleSearch = (query) => {
        if (query) {
            const formattedQuery = query.toLowerCase();
            const filteredData = filter(fullData, (item) => item && contains(item, formattedQuery));
            setData(filteredData);

            // console.log(`Searching for ${query}`);
        } else {
            // If query is cleared, reset the data to the full list
            setData(fullData);
        }
       
    };

    const contains = (item , query) => item.includes(query)
    

    // for handling pressing of location
    const router = useRouter();

    // for pop out of details page
    const pressHandler = async (location) => {
        console.log(location);
        router.push({ pathname: '/PlaceDetailsScreen', params: { locName: location }})
        
    };

    
    
    const [isLoading, setIsLoading] = useState(false);
    const [data,setData] = useState([]);
    const [error, setError] = useState(null);
    const [fullData, setFullData] = useState([]);

    useEffect(() => {
        const fetchData = async () => {
            try {
                const colRef = collection(db,'malls',currentMall,'stores')

                
                // await until data is fetched
                const storeListSnapShot = await getDocs(colRef)

                const storeList = storeListSnapShot.docs.map(doc => doc.id)
                
                console.log(storeList);
                // sorting the list by alphabets
                storeList.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));

                // filter out the unnecessary nodes
                const filteredStoreList = storeList.filter(item => item && !item.toLowerCase().includes('aesop'));

                setFullData(filteredStoreList);

                // set data to full data set initially
                setData(filteredStoreList);

                setIsLoading(false);
                
            } catch (err) {
                console.error("Error fetching data: ", err);
                setError(err);
                setIsLoading(false);
            }
        };
        setIsLoading(true);
        fetchData();
        
    }, [currentMall]);

    
    
    // loading interface
    if (isLoading) {
        return (
            <View style={{flex:1, justifyContent:"center", alignItems:"center" }}>
                <ActivityIndicator size="large" color="#5500dc" />
            </View>
        );
    }

    // error interface
    if ( error ) {
        return (
            <View style={{flex:1, justifyContent:"center", alignItems:"center" }}>
                <Text>Error in fetching data ... Please check your internet connection!</Text>
            </View>
        );
    }

        
    if( currentMall!= null ) {
    return (
        <SafeAreaView style= {styles.mainContainer}>
            <MallPicker
                style ={{flex:0.2,alignItems:'center', justifyContent:'center'}}
                currentMall={currentMall}
                setCurrentMall={setCurrentMall}
                malls={malls}
            />
            <View style = {{ flex:1, paddingBottom: 70}}>
                <Text>Store Search</Text>
                <SearchBar onSearch={handleSearch} />
                
                <FlatList
                data = {data}
                keyExtractor = {(item, index) => index.toString()}
                renderItem = {({item}) => (

                    
                    <TouchableOpacity onPress = { () => pressHandler(item) } >
                    <View style = {styles.itemContainer}>
                        <Image source = {{uri: "https://frameandkeyrealestate.files.wordpress.com/2019/04/clock-icon.png"}}
                        style = {styles.image} />
                        <Text style = {styles.textName}>{item}</Text>
                        {console.log(JSON.stringify(item))}

                    </View>
                    </TouchableOpacity>
                )}
                />
            </View>
        </SafeAreaView>
    );
}
}

const styles = StyleSheet.create({
    mainContainer: {
        flex: 1, 
        marginHorizontal: 20,
    },
    itemContainer: {
        flexDirection: "row",
        alignItems: "center",
        padding: 10,
        marginLeft: 0,
        marginTop: 10,
        borderRadius: 10,
        borderTopWidth: 0.2,
        height:100
    },
    textName: {
        flex: 1,
        fontSize: 17,
        marginLeft: 10,
        fontWeight: "600",

    },
    image:  {
        marginLeft: 10,
        marginTop: 5,
        marginBottom: 5,
        width: 25,
        height: 25,
        borderRadius: 5,
    }
})