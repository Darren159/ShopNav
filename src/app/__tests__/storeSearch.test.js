import { render, fireEvent, waitFor } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import { Alert } from "react-native";
import StoreSearch from "../storeSearch";
import { MallContext } from "../context/mallProvider";
import fetchStoreList from "../../services/fetchStoreList";

jest.mock("../../services/fetchStoreList");

jest.mock("../../services/fetchMalls");

const setCurrentMallMock = jest.fn();

const mockedContextValue = {
  currentMall: "Mall 1",
  setCurrentMall: setCurrentMallMock,
  malls: ["Mall 1", "Mall 2", "Mall 3"],
};

const storeList = [
  { id: "1", name: "Store 1" },
  { id: "2", name: "Store 2" },
];

jest.spyOn(Alert, "alert");

describe("StoreSearch", () => {
  it("renders correctly", async () => {
    const tree = renderer
      .create(
        <MallContext.Provider value={mockedContextValue}>
          <StoreSearch />
        </MallContext.Provider>
      )
      .toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });

  it("renders StoreList with fetched data", async () => {
    fetchStoreList.mockResolvedValue(storeList);
    const { getByText } = render(
      <MallContext.Provider value={mockedContextValue}>
        <StoreSearch />
      </MallContext.Provider>
    );

    await waitFor(() => {
      expect(getByText("Store 1")).toBeTruthy();
      expect(getByText("Store 2")).toBeTruthy();
    });
  });

  it("filters data based on search query", async () => {
    fetchStoreList.mockResolvedValue(storeList);
    const { getByText, getByPlaceholderText, queryByText } = render(
      <MallContext.Provider value={mockedContextValue}>
        <StoreSearch />
      </MallContext.Provider>
    );

    await waitFor(() => {
      fireEvent.changeText(getByPlaceholderText("Search..."), "Store 2");
      expect(getByText("Store 2")).toBeTruthy();
      expect(queryByText("Store 1")).toBeNull();
    });
  });

  it("shows an error alert when fetchStoreList fails", async () => {
    const error = new Error("An error occurred");

    fetchStoreList.mockRejectedValue(error);

    render(
      <MallContext.Provider value={mockedContextValue}>
        <StoreSearch />
      </MallContext.Provider>
    );

    await waitFor(() =>
      expect(Alert.alert).toHaveBeenCalledWith(
        "Error",
        error.message,
        [{ text: "OK" }],
        { cancelable: false }
      )
    );
  });

  it("shows loading indicator while fetching data", async () => {
    fetchStoreList.mockResolvedValue(storeList);
    const { getByTestId, queryByTestId } = render(
      <MallContext.Provider value={mockedContextValue}>
        <StoreSearch />
      </MallContext.Provider>
    );
    expect(getByTestId("loading")).toBeTruthy();

    await waitFor(() => expect(queryByTestId("loading")).toBeNull());
  });
});
