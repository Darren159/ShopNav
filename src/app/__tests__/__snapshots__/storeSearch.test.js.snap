// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreSearch renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
      "marginHorizontal": 20,
    }
  }
>
  <View>
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      clearButtonMode="always"
      onChangeText={[Function]}
      placeholder="Search..."
      style={
        {
          "borderColor": "grey",
          "borderRadius": 10,
          "borderWidth": 1,
          "height": 40,
          "marginTop": 10,
          "paddingLeft": 10,
        }
      }
      value=""
    />
  </View>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
